Milestone 2.5: Lesson Translation Dashboard

DubeySandeep opened the pull request:
- [x] Adds audio translation bar in translation tab.

This PR contains the commits of #5210, So it will get merged after #5210.

DubeySandeep committed:
Added state-translation directive

DubeySandeep committed:
Added state-translation directive

DubeySandeep committed:
Added libraries

DubeySandeep committed:
Added recording bar

DubeySandeep committed:
Added review changes

DubeySandeep committed:
Fixed errors

DubeySandeep committed:
Fixed test failing issues.

DubeySandeep committed:
Added recording bar and player

DubeySandeep committed:
Added timeout for tab refresh

DubeySandeep committed:
Fixed tests fails

DubeySandeep committed:
Added changes to fix tests

DubeySandeep committed:
Removed unwanter broadcast

DubeySandeep committed:
Added Mark needs update functionality

DubeySandeep committed:
Added time filter

DubeySandeep committed:
Updated with develop

DubeySandeep committed:
Resolved lint errors

DubeySandeep committed:
Added warning for record saving

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -0,0 +1,307 @@
+            }
+          };
+
+          $scope.generateNewFilename = function() {
```
I don't think this method needs to be attached to $scope

DubeySandeep replied:
Done!

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -0,0 +1,307 @@
+          $scope.saveRecordedAudio = function() {
+            var filename = $scope.generateNewFilename();
+            var fileType = 'audio/mp3';
+            var recodedAudioFile = new File(
```
recorded*

DubeySandeep replied:
Done!

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -0,0 +1,307 @@
+            AssetsBackendApiService.saveAudio(
+              ExplorationContextService.getExplorationId(), filename,
+              recodedAudioFile).then(function() {
+              if ($scope.isUpdatingAudio) {
```
rename to $scope.audioIsUpdating

DubeySandeep replied:
Done!

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -0,0 +1,307 @@
+                  filename, recodedAudioFile.size);
+              stateContentIdsToAudioTranslationsService.saveDisplayedValue();
+              AlertsService.addSuccessMessage(
+                'Succesfuly uploaded recoded audio.');
```
recorded*

DubeySandeep replied:
Done!

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -0,0 +1,307 @@
+            });
+          };
+
+          $timeout(function(){
```
Why does this need a $timeout?

tjiang11 commented on core/templates/dev/head/pages/header_js_libs.html:
```diff
@@ -14,3 +14,22 @@
+<script src="/third_party/static/wave-surfer-js-1.0.46/wavesurfer.min.js"></script>
+<script src="/third_party/static/angular-recorder-1.4.1/dist/angular-audio-recorder.min.js"></script>
+
+<script>
```
Kind of confused why there is CKEDITOR stuff here?

DubeySandeep replied:
I think this happened because of stacked PR! Removed!

tjiang11 commented on manifest.json:
```diff
@@ -112,6 +112,20 @@
+        "downloadFormat": "zip",
+        "url": "https://github.com/logbon72/angular-recorder/archive/1.4.1.zip",
+        "rootDirPrefix": "ngAudioRecorder-",
+        "targetDirPrefix": "angular-recorder-"
```
Do you need to specify bundle or files here, when I try running the local server it seems it can't find this library

tjiang11 replied:
Nvm about this

tjiang11 reviewed (changes requested):
Thanks @DubeySandeep !
Left a few comments. Currently, I can't test the UI because there seems to be a problem in downloading the third party libraries.

tjiang11 commented:
Comments on UI:

- For the upload-from-computer button, would it make more sense to use a folder icon with the arrow pointing up instead of a plus icon? I feel like that's more common for signifying "Upload". The plus is kind of vague.

- Should the play button be visible at all when there is no audio available? Or should it at least be disabled? Right now it still looks green/clickable.

- Un-capitalize "available" in "No audio Available". Also, I'm trying to think if there's a better phrase for this, because "No audio available" sounds more like something you would say to the learner not the creator. Perhaps "No audio recorded."?

- For the timer, I think it makes more sense to have it count up rather than down, because knowing how long the recording currently is seems like it could be useful for the creator. And maybe you could also do X:XX / 5:00 to show them that 5 minutes is the limit, but I'm not completely sure if it's necessary, since 5 minutes is a very large overestimate of what we expect. It could also perhaps be like, a red countdown somewhere else in the last minute, not sure.

- ![image](https://user-images.githubusercontent.com/11807091/42707754-dce4eb0a-86a0-11e8-97fb-5e60c2ab62c8.png)

Couple things about the above: The timer should not be clickable. The timer should not be cutoff. The green dividers don't look entirely uniform and looks like there's some weird overlap going on that makes it look pretty messy.

- You already are aware of this, but just noting it here for reference. A duplicate waveform should not expand as you are playing back the audio. Rather, only the played back portion of the single waveform should be colored.

- Another issue with the waveform: It doesn't properly adjust as I change the screen. For example, when I zoom out just a little, there's an awkward cutoff at the end of the waveform canvas:

![image](https://user-images.githubusercontent.com/11807091/42708042-a62788c4-86a1-11e8-9456-945e71a64ac2.png)

- I'm not entirely sure if the cloud upload button makes the most sense for the user when they want to save their recording. Yes, we are technically uploading it to GCS in the background, but that's not really important for the user. The user just needs to know that they are confirming what they just recorded. Maybe just a typical "save" button (the floppy disk one, even though floppy disks don't exist anymore that's pretty funny).

- I just tried uploading a file that is bigger than 1MB, and I got an error. This restriction is only present on local dev instances, but the entire bar broke and I can't proceed. It should at least be fail safe and not break when an invalid audio file is encountered, and a useful message should be displayed whenever that happens.

- Can you also include the playback time / total time when playing back recorded audio?

- I think we discussed this a bit before, but having both the re-record and delete buttons after the audio has been saved kind of confuses me. Personally, I would think just having the delete button would make more sense. Because, you are asking for a confirmation when you delete since it's a destructive action, but when you press the re-record button it automatically deletes the recorded audio without warning. So you should either put a confirmation to re-record or just get rid of the re-record button after it has been saved. Also, if a user really wants to re-record it's as easy as deleting and then recording a new audio, which I don't think is bad considering they already confirmed to save the current audio.

- ![image](https://user-images.githubusercontent.com/11807091/42708604-729d2ee4-86a3-11e8-85fe-669eb948ddd4.png) The playback bar should not be overlapping the divider to the left. I think a good deal more padding on both sides would make it look cleaner.

- Should we disallow switching states and/or language code while the user is recording, or after they have finished a draft recording? Let's say a user finishes recording but does not save it, then they switch language or state. Should the recording just be deleted?

DubeySandeep committed:
Updated with develop

anmolshkl commented:
Hi @DubeySandeep, did you get a chance to address Tony's comments?

DubeySandeep commented:
> For the upload-from-computer button, would it make more sense to use a folder icon with the arrow pointing up instead of a plus icon? I feel like that's more common for signifying "Upload". The plus is kind of vague.

I agree with this point and I have replaced wit with upload icon.

> Should the play button be visible at all when there is no audio available? Or should it at least be disabled? Right now it still looks green/clickable.

It was disabled but as I was using div earlier it wasn't working, I have made the changes.

> Un-capitalize "available" in "No audio Available". Also, I'm trying to think if there's a better phrase for this, because "No audio available" sounds more like something you would say to the learner not the creator. Perhaps "No audio recorded."?

Done!

> For the timer, I think it makes more sense to have it count up rather than down, because knowing how long the recording currently is seems like it could be useful for the creator. And maybe you could also do X:XX / 5:00 to show them that 5 minutes is the limit, but I'm not completely sure if it's necessary, since 5 minutes is a very large overestimate of what we expect. It could also perhaps be like, a red countdown somewhere else in the last minute, not sure.
Couple things about the above: The timer should not be clickable. The timer should not be cutoff. The green dividers don't look entirely uniform and looks like there's some weird overlap going on that makes it look pretty messy.

Done! 

> You already are aware of this, but just noting it here for reference. A duplicate waveform should not expand as you are playing back the audio. Rather, only the played back portion of the single waveform should be colored.

That was occuring because of css, It's resolved now!

> Another issue with the waveform: It doesn't properly adjust as I change the screen. For example, when I zoom out just a little, there's an awkward cutoff at the end of the waveform canvas:

Not sure how to handel this and I'm not aware why someone will  zoom page while recording (although it works fine if you'll zoom and the start recording)

> I'm not entirely sure if the cloud upload button makes the most sense for the user when they want to save their recording. Yes, we are technically uploading it to GCS in the background, but that's not really important for the user. The user just needs to know that they are confirming what they just recorded. Maybe just a typical "save" button (the floppy disk one, even though floppy disks don't exist anymore that's pretty funny).

I did this because we already have a save draft button at the top it doesn't make sense to save audio once and the again save draft! (Although i have replaced it with save(floppy) icon).

> I just tried uploading a file that is bigger than 1MB, and I got an error. This restriction is only present on local dev instances, but the entire bar broke and I can't proceed. It should at least be fail safe and not break when an invalid audio file is encountered, and a useful message should be displayed whenever that happens.

I cannot see anything breaking here, I have added a AlertService.warning for faliures!

> I think we discussed this a bit before, but having both the re-record and delete buttons after the audio has been saved kind of confuses me. Personally, I would think just having the delete button would make more sense. Because, you are asking for a confirmation when you delete since it's a destructive action, but when you press the re-record button it automatically deletes the recorded audio without warning. So you should either put a confirmation to re-record or just get rid of the re-record button after it has been saved. Also, if a user really wants to re-record it's as easy as deleting and then recording a new audio, which I don't think is bad considering they already confirmed to save the current audio.

When you press re-recoding button nothing gets deleted you'll recorde once and if you'll "save" then only the old audio file will get replaced by the new one. Should I add a similar warning modal for this? 

> The playback bar should not be overlapping the divider to the left. I think a good deal more padding on both sides would make it look cleaner.

Done!

> Should we disallow switching states and/or language code while the user is recording, or after they have finished a draft recording? Let's say a user finishes recording but does not save it, then they switch language or state. Should the recording just be deleted?

Yes, as they haven't saved their work (I can see similar funcionality on editor page, edit state content and switch to other tab without saving your work, ans now go back to editor tab)

DubeySandeep committed:
Added review changes

vojtechjelinek commented on manifest.json:
```diff
@@ -112,6 +112,13 @@
+        "url": "https://github.com/logbon72/angular-recorder/archive/1.4.1.zip",
+        "rootDirPrefix": "angular-recorder-",
+        "targetDirPrefix": "angular-recorder-"
+      },
```
Do you need more files from this zip other than *angular-audio-recorder.min.js*? If you don't, then try to way to download just *angular-audio-recorder.min.js* instead of the whole zip.

DubeySandeep replied:
@vojtechjelinek, The build file needs lame.js file which is in ../lib dir so I'll have to download the zip file otherwise the mp3 converter will not work.

Although they have the option to import lame.js through URL but that's not working right now ( see 3rd bulletin [here](https://github.com/sathyapulse/angular-recorder#provider-recorderserviceprovider)). I have talked with Tony regarding this we have to go this way for now. Also, this zip contains at most 30 files.

vojtechjelinek replied:
Ah, ok, just to check how many files are in the unzipped folder?

DubeySandeep replied:
30 files (approx.)

vojtechjelinek replied:
Ok, I think it is okay to keep it as zip for now.

vojtechjelinek reviewed (changes requested):
PTAL

anmolshkl commented:
Hi @DubeySandeep , extremely sorry for the late review. I took a look and noticed the following things:

- The audio  recorder is working fine on Firefox but not on Chrome (build 67.x on Ubuntu 16.04). When I click on the record button, the timer starts but nothing gets recorded and the audio-waveform is empty too. I checked the microphone permission, but the microphone access was given to the page.
- The box that contains the audio-waveform, expanded when I clicked on recording button hence going out of bounds.
![audio-player-issue-3-mz](https://user-images.githubusercontent.com/6338799/43217407-2d08e646-905f-11e8-94f5-a24d24d7a031.png)
- Although the page gets loaded immediately, the audio recorder takes some time (20s?) to load (it keeps showing "Loading..." for around 20-30s). I'm using Chrome 67.x on Ubuntu 16.04. Now, many applications run really slowly on my Ubuntu due to some hardware compatibility issues of Ubuntu with MSI. I'm not sure if this behavior is because of this reason or not. Could you please confirm with someone else (perhaps other students?) and see if this issue is common?
- On Firefox, the recording worked fine, but, there were some other issues that I observed.
![audio-player-issue-2-mz](https://user-images.githubusercontent.com/6338799/43217272-d13acc58-905e-11e8-8206-d2f495158a8f.png)
    - After recording the audio, the waveform shrinks so much so that it is almost invisible.
    - The vertical line in the audio-waveform which marks the progress of the audio, stops at 99% mark when the audio playback gets over.

I'm not sure if the Chrome issue is something one-off or something that might pop-up on other people's laptop/browser too. I'd recommend that you ask other students to try out the audio player on their browser and see what issues they face. This could be one possible way of doing user testing on a small scale. Meanwhile, I also recommend pushing the PR for Milestone 3.2 (since it is independent).

PS: I'll try to see what could be wrong with my Chrome browser, and why it is not recording anything.

tjiang11 commented:
Hi Sandeep,

This is looking a lot better than before. Some more comments:

- Maybe the tooltip for the "Upload to cloud" button should be "Save" instead of "Upload", because "Upload" is being used as the tooltip text for the "Upload from computer" button as well, so it may be confusing to have the same tooltip text for two different buttons.

- After recording and saving, I think the starting point of the playback bar is still too close to the left border. The draggable circle/thumb should not be touching the left border separating it and the play button. 

- After recording and saving, I think there should be a time when playing back audio, in the format X:XX/X:XX (playback time / total time), perhaps to the left or above the playback bar.

- After recording but not saving, during playback, the cursor should reset to the beginning after the whole audio has been played through.

And to give another machine's perspective on Anmol's comments:

(Ubuntu 16.04, Chrome 61.0.3163.79)

- The waveform size/magnitude is actually pretty reasonable on my machine. It's unclear if Anmol's recording is actually much softer than normal.
- The audio recorder loads pretty quickly for me. I don't notice any delay between it and the rest of the page.
- I'm able to record.

DubeySandeep committed:
Added review changes

DubeySandeep committed:
Added review changes(2)

DubeySandeep committed:
Updated with upstream/develop

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -242,14 +250,17 @@ oppia.directive('audioTranslationBar', [
                 languageCode: function() {
                   return $scope.languageCode;
+                },
+                isUpdatingAudio: function() {
```
This name is pretty misleading: at first glance, I thought this meant that some audio file was in the process of updating or something. Maybe re-name to something along the lines of "isAudioAvailable"?

DubeySandeep replied:
Done!

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/add_audio_translation_modal_directive.html:
```diff
@@ -13,6 +13,10 @@ <h3>Add Audio Translation</h3>
             </audio-file-uploader>
+            <div ng-if="isUpdatingAudio"
+                 class="oppia-updated-audio-file-upload-message">
+              <strong>Note:</strong> You're going to update the audio translation and this will delete the already available audio.
```
Maybe make it more concise, something like: `Uploading a new audio file will replace the existing audio file.`

DubeySandeep replied:
Done!

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/add_audio_translation_modal_directive.html:
```diff
@@ -13,6 +13,10 @@ <h3>Add Audio Translation</h3>
                                  on-file-changed="updateUploadedFile"
                                  on-file-cleared="clearUploadedFile">
             </audio-file-uploader>
+            <div ng-if="isUpdatingAudio"
```
I think it might be better if this message goes between the license notice and the upload button.

tjiang11 replied:
Nevermind, this seems like it could be a hassle since the license notice is in the directive. Up to you

DubeySandeep replied:
I think we can leave it as-is!

tjiang11 reviewed (changes requested):
- Does it make sense for the audio for main state content to be able to be recorded when there is no content yet?

- After recording and playing back the waveform preview, the audio continues to play even after pressing cancel.

- Let's say a user comes in to record, and they inadvertently click the "Block" button when they are asked for microphone permissions. There is no way for them to retry the permissions unless they refresh the page, since the record button is now disabled. Is this okay?

- After recording and saving, I press play and let the audio play once. The thumb resets to the beginning of the recording. I press play again, and nothing happens.

- Let's say I record an audio file that's 2-3 minutes long, so it takes a while to process into an mp3 file by the browser. I click on a different audio tab to record something else while it's uploading. Oops! My 2-3 minute recording is gone. Should we prevent this? In fact, this happens also when the processing is done and the audio preview is shown, which can be even more misleading. Maybe disable outward navigation while recording, saving, and previewing?

DubeySandeep commented:
> Does it make sense for the audio for main state content to be able to be recorded when there is no content yet?

Translation tab is made for translators, I think any exploration owner will assign a translator only after completing his editing work, so maybe we can leave that as is! 

> After recording and playing back the waveform preview, the audio continues to play even after pressing cancel.

Fixed!

> Let's say a user comes into record, and they inadvertently click the "Block" button when they are asked for microphone permissions. There is no way for them to retry the permissions unless they refresh the page, since the record button is now disabled. Is this okay?

That's an intended behaviour of the browser!

> After recording and saving, I press play and let the audio play once. The thumb resets to the beginning of the recording. I press play again, and nothing happens.

Fixed!

> Let's say I record an audio file that's 2-3 minutes long, so it takes a while to process into an mp3 file by the browser. I click on a different audio tab to record something else while it's uploading. Oops! My 2-3 minute recording is gone. Should we prevent this? In fact, this happens also when the processing is done and the audio preview is shown, which can be even more misleading. Maybe disable outward navigation while recording, saving, and previewing?

I have added a warning message for this currently we cannot disable the whole thing! and in the worst case, (where we can have design issues) I've added loadingMessage!

DubeySandeep committed:
Added review changes

DubeySandeep committed:
Added audio load progress in correct position

tjiang11 commented:
I found a bug, that I don't think was there before:
- Make some content and a feedback in the editor tab.
- Go to the translation tab
- Start recording audio for the main content
- While still recording, click on the Feedback tab, and the page breaks.

Also, i see that you've added in the "record new" button. I originally advocated for not having this button, but am not completely against it either. Anyways, with the current design, something about the flow doesn't make sense:
- Record audio for the main content, and save it.
- I click on the "record new audio" button.
- I listen to the preview, and decide to keep the old one since I like it better. I press the "cancel" button. Now there's no audio, but the original audio should be there.

Also, I don't know if you've noticed, but in the audio waveform preview, dragging the cursor has no effect whatsoever.

tjiang11 commented:
Also, I'm seeing a bunch of these errors pop up when switching between states in the translation tab.
![image](https://user-images.githubusercontent.com/11807091/43657716-38ee5068-971c-11e8-9731-2ac9312beade.png)

To reproduce create a new state and immediately navigate to it in the translation tab.

DubeySandeep committed:
Fixed tab change issue

tjiang11 commented:
Bug:
- Start recording audio for content
- While recording, switch to feedback tab
- Switch back to content tab and start recording again.
- Look at the timer. It's going up two seconds for every second while recording.

DubeySandeep committed:
Fixed state change issue

DubeySandeep committed:
Fixed timer issue

DubeySandeep commented:
@tjiang11, Fixed all issue!
Also cannot reproduce this one:
> Also, i see that you've added in the "record new" button. I originally advocated for not having this button, but am not completely against it either. Anyways, with the current design, something about the flow doesn't make sense:
Record audio for the main content, and save it.
I click on the "record new audio" button.
I listen to the preview, and decide to keep the old one since I like it better. I press the "cancel" button. Now there's no audio, but the original audio should be there.

tjiang11 commented:
Noticed another weird bug:
- Record an audio for an answer group.
- Press save button
- WHILE it is saving, switch to another answer group.
- The audio is saved under the new answer group!

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/audio_translation_bar_directive.html:
```diff
@@ -0,0 +1,227 @@
+
+<div class="oppia-translation-bottom-right-container" ng-if="showRecorderWarning">
+  <span>
+    <strong>Warning: </strong>Please keep this tab open until you save the recorded audio. If you navigate elsewhere before that, the recorded audio will be lost.
```
So upon reading this, my instinct is to think Chrome tab, so that as long as I have the current browser tab open, the audio will not be lost. I'm not sure if there's a way to make this more clear/explicit? Effectively, the user shouldn't do ANYTHING else while it's being recorded, converted, previewed, and saved. The main concern is conversion because that can take a long time for longer files, so the user should just sit there and not touch anything.

DubeySandeep committed:
buggy file save fixed

tjiang11 commented:
Design:
- After saving audio, it always say 0:00/0:00 while the audio is not playing, both before starting, and after the audio has finished playing. I think we should either display 0:00/X:XX (where X:XX is the total time) OR we just don't show the timer while it's not playing. 0:00/0:00 is misleading information.

tjiang11 commented:
Error:
- Start recording feedback in one state, and while recording switch states. I get this and the entire page **breaks:**
![image](https://user-images.githubusercontent.com/11807091/43660989-35ce741c-9726-11e8-8468-eed13a4acf76.png)

DubeySandeep committed:
Fixed feedback to state change issue

DubeySandeep committed:
Changed timer design

tjiang11 commented:
Bug:
- The timer incrementing too quickly also occurs when starting the recording and switching the language code.

DubeySandeep committed:
Fixed lang change issue

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -225,14 +225,14 @@ oppia.directive('audioTranslationBar', [
           };
 
           $scope.getUploadedAudioTrimer = function() {
```
getUploadedAudioTimer*

tjiang11 commented:
Did a final pass and I think it looks good

DubeySandeep commented:
Hi @tjiang11  I've resolved the merge conflicts and added the functionality to pop-up a modal when users will click other buttons of the translation tab while recording!
The commit is going through lint checks please wait!

DubeySandeep committed:
Added pop-up funcionality when navigating while recording

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -207,6 +223,34 @@ oppia.directive('audioTranslationBar', [
+          $scope.openTranslationTabBusyModal = function() {
+            $uibModal.open({
+              templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+                '/pages/exploration_editor/translation_tab/' +
```
Where is this template defined?

DubeySandeep committed:
Added modal html file

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -84,12 +85,27 @@ oppia.directive('audioTranslationBar', [
+            var message = '';
+            if($scope.recorder.status.isRecording) {
+              message = 'It seems like you haven\'t stopped recording, please' +
+                ' stop recording and either save or cancel the recording!'
```
Use a period instead of an exclamation point, ditto for the rest. Also, make it two separate sentences. The first sentence could also be more concise. So maybe something like this:

`You haven't finished recording. Please stop recording and either save or cancel the recording.`

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -84,12 +85,27 @@ oppia.directive('audioTranslationBar', [
+                ' stop recording and either save or cancel the recording!'
+            } else if($scope.recorder.status.isConverting) {
+              message = 'It seems like recorded audio is still getting ' +
+                'converted into mp3 please wait till it\'s done!'
```
Maybe: `Please wait until the audio has finished processing.`

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -84,12 +85,27 @@ oppia.directive('audioTranslationBar', [
+                'converted into mp3 please wait till it\'s done!'
+            } else if($scope.showRecorderWarning) {
+              message = 'It looks like you haven\'t saved the recorded audio,' +
+                ' please save/cancel the recorded audio.'
```
Maybe: `You haven't saved your recording. Please save or cancel the recording.`

tjiang11 commented:
Bug:
- Start recording something for the main content.
- While recording attempt to switch to the feedback tab. The modal opens up preventing you from switching. Press "Go it"
- Now, finish recording, and attempt to save the audio. It errors.

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/translation_tab/translation_tab_busy_modal_directive.html:
```diff
@@ -0,0 +1,9 @@
+<div class="modal-header">
+  <h4 class="modal-title">Roecording process in progress!</h4>
```
Recording*

And remove process.

So `Recording in progress!`

tjiang11 replied:
Or perhaps `Finish your recording`

tjiang11 commented:
Not sure why but when I load the translation tab the language is set to Hebrew by default, is it the same for you?

DubeySandeep committed:
Added review changes

tjiang11 commented on scripts/pre_commit_linter.py:
```diff
@@ -420,7 +420,7 @@ def _lint_js_files(
 
     print 'Total js files: ', num_js_files
-    eslint_cmd_args = [node_path, eslint_path, '--quiet']
+    eslint_cmd_args = [node_path, eslint_path, '--quiet', '--fix']
```
What does this do?

DubeySandeep committed:
Fixed lint errors

tjiang11 commented:
WIll merge after tests pass.

DubeySandeep committed:
Merge branch 'develop' of https://github.com/oppia/oppia into Milestone_2.5

DubeySandeep committed:
Fixed merge conflicts
